### PR TITLE
Convert mixin stuff to use asm

### PR DESCRIPTION
**Pull request by Earthcomputer**
Convert mcdev's mixin features to use ASM instead of PSI for the target class. Solves some long-standing issues with mcdev mixin support, and paves the way for even better support in the future.

### CODE REVIEW

**Comment by Earthcomputer** [comment]
This is now working, and it's a solution I'm mostly happy with. Most of the remaining work is just code cleanup and documentation

**Comment by Earthcomputer** [comment]
This is now finished and ready for review.

File: build.gradle.kts [hunk 1]
```diff
@@ -108,7 +109,8 @@ intellij {
         "org.toml.lang:$pluginTomlVersion",
         // needed dependencies for unit tests
         "properties",
-        "junit"
+        "junit",
+        "ByteCodeViewer"
```
Comment (RedNesto) [inline build.gradle.kts]: Is this plugin needed for anything ?

File: src/main/kotlin/platform/mixin/insight/MixinTargetLineMarkerProvider.kt [hunk 2]
```diff
@@ -66,8 +66,18 @@ class MixinTargetLineMarkerProvider : LineMarkerProviderDescriptor() {
...
+            it,
+            "shadow"
+        )
+    }
+    private fun PsiMember.findAccessor(): TargetInfo? = this.findAccessorTargetForReference()?.let {
+        TargetInfo(
+            it,
+            "accessor"
+        )
+    }
```
Comment (RedNesto) [inline src/main/kotlin/platform/mixin/insight/MixinTargetLineMarkerProvider.kt]: Minor nitpick.
```suggestion
    private fun PsiMember.findShadow(): TargetInfo? =
        this.findFirstShadowTargetForReference()?.let { TargetInfo(it, "shadow") }

    private fun PsiMember.findAccessor(): TargetInfo? =
        this.findAccessorTargetForReference()?.let { TargetInfo(it, "accessor") }

```

File: src/main/kotlin/platform/mixin/inspection/injector/InjectorType.kt [hunk 3]
```diff
@@ -160,32 +186,44 @@ enum class InjectorType(private val annotation: String) {
             val parameters = ArrayList<Parameter>(2)
 
             // TODO: Report if GETSTATIC/PUTSTATIC is used for an instance field
-            if (opcode == Opcodes.GETFIELD || opcode == Opcodes.PUTFIELD) {
-                parameters.add(getInstanceParameter(owner))
+            if (!field.hasAccess(Opcodes.ACC_STATIC)) {
+                parameters += Parameter(
+                    "self",
+                    elementFactory.createTypeFromText(Type.getType(ownerName.replace('.', '/')).className, null)
```
Comment (RedNesto) [inline src/main/kotlin/platform/mixin/inspection/injector/InjectorType.kt]: This causes the following exception in some mixins (like [Sponge's MinecraftMixin](https://github.com/SpongePowered/Sponge/blob/fecd4239bed9ca5b769d700b32c858e4d3ff09cc/src/mixins/java/org/spongepowered/common/mixin/core/client/MinecraftMixin.java))

```
java.lang.IllegalArgumentException
	at org.objectweb.asm.Type.getTypeInternal(Type.java:440)
	at org.objectweb.asm.Type.getType(Type.java:177)
	at com.demonwav.mcdev.platform.mixin.inspection.injector.InjectorType$REDIRECT.collectFieldParameters(InjectorType.kt:192)
	at com.demonwav.mcdev.platform.mixin.inspection.injector.InjectorType$REDIRECT.expectedMethodSignature(InjectorType.kt:125)
	at com.demonwav.mcdev.platform.mixin.inspection.injector.InvalidInjectorMethodSignatureInspection$Visitor.visitMethod(InvalidInjectorMethodSignatureInspection.kt:78)
	at com.intellij.psi.impl.source.PsiMethodImpl.accept(PsiMethodImpl.java:284)
	at com.intellij.codeInspection.InspectionEngine.acceptElements(InspectionEngine.java:66)
	at com.intellij.codeInspection.InspectionEngine.createVisitorAndAcceptElements(InspectionEngine.java:57)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.runToolOnElements(LocalInspectionsPass.java:298)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.lambda$visitPriorityElementsAndInit$3(LocalInspectionsPass.java:267)
	at com.intellij.util.AstLoadingFilter.forceAllowTreeLoading(AstLoadingFilter.java:156)
	at com.intellij.util.AstLoadingFilter.forceAllowTreeLoading(AstLoadingFilter.java:148)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.lambda$visitPriorityElementsAndInit$4(LocalInspectionsPass.java:266)
	at com.intellij.util.AstLoadingFilter.disallowTreeLoading(AstLoadingFilter.java:127)
	at com.intellij.util.AstLoadingFilter.disallowTreeLoading(AstLoadingFilter.java:116)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.lambda$visitPriorityElementsAndInit$5(LocalInspectionsPass.java:266)
	at com.intellij.concurrency.ApplierCompleter.execAndForkSubTasks(ApplierCompleter.java:136)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1137)
	at com.intellij.concurrency.ApplierCompleter.lambda$wrapInReadActionAndIndicator$1(ApplierCompleter.java:92)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:658)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:610)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:65)
	at com.intellij.concurrency.ApplierCompleter.wrapInReadActionAndIndicator(ApplierCompleter.java:104)
	at com.intellij.concurrency.ApplierCompleter.lambda$compute$0(ApplierCompleter.java:83)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:172)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:183)
	at com.intellij.concurrency.ApplierCompleter.compute(ApplierCompleter.java:83)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

File: src/main/kotlin/platform/mixin/inspection/injector/InjectorType.kt [hunk 4]
```diff
@@ -109,48 +115,68 @@ enum class InjectorType(private val annotation: String) {
...
+        ): Pair<List<Parameter>, PsiType>? {
+            val elementFactory = JavaPsiFacade.getElementFactory(project)
+            val ownerName = reference.owner ?: return null
+
+            val hasThis = !method.hasAccess(Opcodes.ACC_STATIC)
+            val parameters = mutableListOf<Parameter>()
+            if (hasThis) {
+                parameters += Parameter(
+                    "self",
+                    elementFactory.createTypeFromText(Type.getType(ownerName.replace('.', '/')).className, null)
```
Comment (RedNesto) [inline src/main/kotlin/platform/mixin/inspection/injector/InjectorType.kt]: This causes the following exception in [Sponge's MobMixin](https://github.com/SpongePowered/Sponge/blob/a65c51431db25a528fafcc4dc89abe55f0bb78b2/src/mixins/java/org/spongepowered/common/mixin/core/world/entity/MobMixin.java)
```
java.lang.IllegalArgumentException
	at org.objectweb.asm.Type.getTypeInternal(Type.java:440)
	at org.objectweb.asm.Type.getType(Type.java:177)
	at com.demonwav.mcdev.platform.mixin.inspection.injector.InjectorType$REDIRECT.collectMethodParameters(InjectorType.kt:157)
	at com.demonwav.mcdev.platform.mixin.inspection.injector.InjectorType$REDIRECT.expectedMethodSignature(InjectorType.kt:120)
	at com.demonwav.mcdev.platform.mixin.inspection.injector.InvalidInjectorMethodSignatureInspection$Visitor.visitMethod(InvalidInjectorMethodSignatureInspection.kt:78)
	at com.intellij.psi.impl.source.PsiMethodImpl.accept(PsiMethodImpl.java:284)
	at com.intellij.codeInspection.InspectionEngine.acceptElements(InspectionEngine.java:66)
	at com.intellij.codeInspection.InspectionEngine.createVisitorAndAcceptElements(InspectionEngine.java:57)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.runToolOnElements(LocalInspectionsPass.java:298)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.lambda$visitPriorityElementsAndInit$3(LocalInspectionsPass.java:267)
	at com.intellij.util.AstLoadingFilter.forceAllowTreeLoading(AstLoadingFilter.java:156)
	at com.intellij.util.AstLoadingFilter.forceAllowTreeLoading(AstLoadingFilter.java:148)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.lambda$visitPriorityElementsAndInit$4(LocalInspectionsPass.java:266)
	at com.intellij.util.AstLoadingFilter.disallowTreeLoading(AstLoadingFilter.java:127)
	at com.intellij.util.AstLoadingFilter.disallowTreeLoading(AstLoadingFilter.java:116)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.lambda$visitPriorityElementsAndInit$5(LocalInspectionsPass.java:266)
	at com.intellij.concurrency.ApplierCompleter.execAndForkSubTasks(ApplierCompleter.java:136)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1137)
	at com.intellij.concurrency.ApplierCompleter.lambda$wrapInReadActionAndIndicator$1(ApplierCompleter.java:92)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:658)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:610)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:65)
	at com.intellij.concurrency.ApplierCompleter.wrapInReadActionAndIndicator(ApplierCompleter.java:104)
	at com.intellij.concurrency.ApplierCompleter.lambda$compute$0(ApplierCompleter.java:83)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:172)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:183)
	at com.intellij.concurrency.ApplierCompleter.compute(ApplierCompleter.java:83)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

File: src/main/kotlin/util/utils.kt [hunk 5]
```diff
@@ -239,3 +296,58 @@ inline fun <reified T> Iterable<*>.firstOfType(): T? {
 }
 
 fun libraryKind(id: String): LibraryKind = LibraryKind.findById(id) ?: LibraryKind.create(id)
+
+fun String.isJavaKeyword() = javaKeywords.contains(this)
```
Comment (RedNesto) [inline src/main/kotlin/util/utils.kt]: I think this can be replaced with `com.intellij.lang.java.lexer.JavaLexer#isKeyword`.

**Review by RedNesto — CHANGES_REQUESTED** [review]
This looks very good.

From my limited tests I found 2 similar exceptions simply when browsing Sponge's mixins.

File: build.gradle.kts (hunk 1 above)
Comment (Earthcomputer) [reply build.gradle.kts]: Yes - for retrieving the bytecode of a class, [here](https://github.com/Earthcomputer/MinecraftDev/blob/0a5892317924abd61a57ec5b84cb753146143278/src/main/kotlin/platform/mixin/util/AsmUtil.kt#L119)

File: src/main/kotlin/platform/mixin/insight/MixinTargetLineMarkerProvider.kt (hunk 2 above)
Comment (Earthcomputer) [reply src/main/kotlin/platform/mixin/insight/MixinTargetLineMarkerProvider.kt]: Thanks ktlintFormat

File: src/main/kotlin/util/utils.kt (hunk 5 above)
Comment (Earthcomputer) [reply src/main/kotlin/util/utils.kt]: Ahh knew there must have been something

**Comment by Earthcomputer** [comment]
Those exceptions were a bit of a silly mistake on my part - fixed now.

File: build.gradle.kts (hunk 1 above)
Comment (RedNesto) [reply build.gradle.kts]: Oh right, could you move it above the comment then, since it's not used for unit tests.

**Comment by RedNesto** [comment]
Found another 2 issues when messing in [Sponge's BlockBehaviour_PropertiesMixin](https://github.com/SpongePowered/Sponge/blob/aab3bddf5a3cf30775a5d88c34ff4be4ef66f488/src/mixins/java/org/spongepowered/common/mixin/core/world/level/block/state/BlockBehaviour_PropertiesMixin.java)

Here's an injection with its suggested method parameters:

```java
@Inject(method = "of(Lnet/minecraft/world/level/material/Material;Ljava/util/function/Function;)Lnet/minecraft/world/level/block/state/BlockBehaviour$Properties;", at = @At("HEAD"))
private static void test(Material param0, Function<BlockState<MaterialColor>> param1, CallbackInfoReturnable<BlockBehaviour.Properties> cir) {
}
```

but the expected method parameters are
```java
@Inject(method = "of(Lnet/minecraft/world/level/material/Material;Ljava/util/function/Function;)Lnet/minecraft/world/level/block/state/BlockBehaviour$Properties;", at = @At("HEAD"))
private static void test(Material param0, Function<BlockState, MaterialColor> param1, CallbackInfoReturnable<BlockBehaviour.Properties> cir) {
}
```

Note the inspection does not report correct parameters, only the fix is wrong in this case.

---

In the same file, breaking the parameters of `impl$onCreate` and using the fix also generates invalid parameters:
```java
private static void impl$onCreate(Material param0, DyeColor param1, CallbackInfoReturnable<net.minecraft.world.level.block.state.BlockBehaviour$Properties> cir) {
}
```

**Comment by Earthcomputer** [comment]
Fixed, and added tests for these cases.

**Review by RedNesto — APPROVED** [review]

**Pull request merged**